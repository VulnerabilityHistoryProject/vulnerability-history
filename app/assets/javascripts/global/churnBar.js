export function churnBar(insertions, deletions) {
  const i_blocks = Math.log10(insertions + 1)
  const d_blocks = Math.log10(deletions + 1)
  
  const i_part = (insertions <= 0) ? '' : `
    <span class="insertions n">+${insertions}</span>
    <span class="insertions-bg bar" style="width: ${i_blocks}em"></span> `
  const d_part = (deletions <= 0) ? '' : `
    <span class="deletions n">+${deletions}</span>
    <span class="deletions-bg bar" style="width: ${d_blocks}em"></span> `
  
  return `
    <span class="churn-bar">
      <span class="num-bar-container">${i_part}</span>
      <span class="num-bar-container">${d_part}</span>
    </span>`
}
