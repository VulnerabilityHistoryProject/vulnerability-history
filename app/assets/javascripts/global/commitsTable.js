/*
 Generic table to be used for commits
 Lists the commit no., commit message, and commit date
 Currently being used on filepaths/show, but should be reused as needed
 */

function buildTopNav() {
    $('#datatable_wrapper').prepend('<div id="table-nav-top"></div>');
    const nav = $('#table-nav-top');
    nav.addClass('table-nav grid-x');

    const nav_length = $('#datatable_length');
    nav_length.addClass('table-nav-show-entries cell small-9')
    nav.append(nav_length);

    const nav_search = $('#datatable_filter label');
    nav_search.addClass('table-nav-search cell small-3')
    nav.append(nav_search);
}

function languageInfo() {
    return {
        search: "", // nothing in the placeholder;
        searchPlaceholder: 'Search',
        lengthMenu: "Show _MENU_",
        info: "Showing _START_ to _END_ of _TOTAL_ commits",
        infoEmpty: "Showing 0 to 0 of 0 entries",
        infoFiltered: "(filtered from _MAX_ total commits)",
    };
}

function editsTable(jsonData, tableName = '#edit_table') {
    $(tableName).dataTable({
        responsive: true,
        destroy: true,
        data: jsonData,
        order: [[2, "desc"]], /* Default sort */
        columnDefs: [],
        columns: [
            {
                title: 'Commit No',
                data: 'commit_hash',
                defaultContent: '',
                className: 'title',
                responsivePriority: 1
            },
            {
                title: 'Commit Message',
                data: 'message',
                defaultContent: '',
                responsivePriority: 2
            },
            {
                title: 'Date',
                data: 'date_created',
                defaultContent: '',
                responsivePriority: 3
            }
            ],
        language: languageInfo(),
        initComplete: function () {
            $('#loading').remove();
            buildTopNav();
        }
    });

    $('#edit_table').on('mouseover', 'tbody tr', function () {
        $(this).css({
            'cursor': 'pointer',
            'color': 'blue'
        });
    })

    $('#edit_table').on('mouseout', 'tbody tr', function () {
        $(this).css({
            'cursor': 'default',
            'color': 'inherit'
        });
    });
};
