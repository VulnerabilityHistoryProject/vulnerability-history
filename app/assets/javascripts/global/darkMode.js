const toggleBtn = document.getElementById("themeToggle");

function prefersLightTheme(){
	return window.matchMedia("(prefers-color-scheme: light)").matches;
}

function setThemeDark(isDark){
	document.body.classList.add(isDark ? 'dark-mode' : 'light-mode');
	document.body.classList.remove(isDark ?  'light-mode' : 'dark-mode');
	toggleBtn.checked = isDark;
	localStorage.setItem("use-dark-mode", isDark)
}

function currentThemeIsDark(){
	return document.body.classList.contains("dark-mode")
}

function decideUseDark(){
	switch(localStorage.getItem("use-dark-mode")){
		case "true":
			return true
		case "false":
			return false
		default:
			return prefersLightTheme()
			// if you have zero preferences and history... DARK MODE!
			// ...because Andy likes VHP dark mode better.
	}
}

/*
  How we determine light/dark mode:

  Precedence:
	1. localStorage
	2. perfers-color-scheme
	3. Dark mode

  Then: store in localStorage after determine from precedence
 */
export default function determineLightDarkMode() {
	const useDarkMode = decideUseDark();
	toggleBtn.addEventListener("click",() => setThemeDark(!currentThemeIsDark()));
	setThemeDark(useDarkMode)
}