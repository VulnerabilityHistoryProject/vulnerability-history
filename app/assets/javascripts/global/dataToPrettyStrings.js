/*
This file is a small utility for formatting data coming from the VHP API
It's mostly just string concatentation with some minor logic.
So that we keep things consistent.
*/

export function vulnProject(data){
    return `<img class="${data.subdomain}-inline-logo inline-logo"
                 title="${data.project_name}"/>`
}

export function vulnName(data) {
  let str = data.cve;
  if (data.nickname != "" && data.nickname != null) {
    str = `<span class="nicknamed-cve"> ${data.nickname}</span>
           <span class="nicknamed-aka-cve"> aka ${data.cve}</span>`;
  }
  return str;
}

export function vulnLink(data, str) {
  return `<a data-id="${data.id}" class="dt_row" href="/${data.id}">${str}</a>`
}

//Simplifies the CWE name if it has a nickname in parentheses
export function friendlyCWE(name) {
  // "I do not want this ('I want this') not this" --> "I want this"
  // "Just a string" --> "Just a string"
  const match = name.match(/.*\(\'(.+)\'\).*/);
  if (match) {
    return match[1]; //[0] is the full string, [1] is the actual match
  } else {
    return name;
  }
}

export function ellipsizedFilepath(fp, prefixLen, maxLen) {
  /* Explanation of this magical regex
     Group 1: from the beginning of the string up to prefixLen, greedily get
              the biggest string you can ending in a /
              e.g. foo/bar/baz --> foo/ for prefixLen=3
     Group 2: at least one character must be here to be ellipsized/discarded,
              again note that we're greedy here so we ellipsize as much as
              possible before getting to group 3.
     Group 3: from the end of the string, get the smallest possible string
              starting with a / (i.e. filename)
  */
  const regex = new RegExp(`^(.{1,${prefixLen}}/)(.+)(/.+)$`, "g");
  const replacer = function (str, prefix, _mid, suffix) {
    if (str.length > maxLen) {
      return `${prefix}...${suffix}`;
    } else {
      return str;
    }
  };
  return [replacer, fp.replace(regex, replacer)];
}
