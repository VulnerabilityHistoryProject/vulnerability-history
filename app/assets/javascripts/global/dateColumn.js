function renderDateColumn(data, type, row, displayFormat, isLink) {
  if (type === 'display') {
    // Display dates in a nice format.
    if (isLink) {
      return renderAsLink(moment(data).format(displayFormat), row);
    } else {
      return moment(data).format(displayFormat);
    }
  } else if (type === 'filter') {
    // Allow for the filtering of dates in a variety of formats.
    // The type is "filter" when the search bar is used.
    let currentDate = moment(data);
    let longDisplay = currentDate.format(displayFormat);
    let shortDisplay = currentDate.format("MM/DD/YYYY");
    return data + " " + longDisplay + " " + shortDisplay;
  } else {
    // Sort dates without any special formatting.
    return data;
  }
}
