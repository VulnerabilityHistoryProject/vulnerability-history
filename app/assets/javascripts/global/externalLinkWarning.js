$(document).ready(function(){
    var root = new RegExp(location.host);

    $('a').each(function(){

        if(root.test($(this).attr('href'))){
            $(this).addClass('local');
        }
        else{
            // a link that does not contain the current host
            var url = $(this).attr('href');
            if(url) {
                if(url.length > 1)
                {
                    $(this).addClass('external');
                }
            }
        }
    });

    $('a.external').on('click', function(e){

        e.preventDefault();
        var answer = confirm("You are about to leave the Vulnerability History Project website and view the content of an external website. Do you want to proceed?");

        if (answer){
            window.location = $(this).attr('href');
        }

    });

});
