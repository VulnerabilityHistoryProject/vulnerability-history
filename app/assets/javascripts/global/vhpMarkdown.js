import showdown from "showdown";
import sanitizeHtml from "sanitize-html";

export function vhpMarkdownAll() {
  for (let e of document.querySelectorAll(".vhp-markdown")) {
    e.innerHTML = vhpMarkdown(e.textContent.trim());
  }
}

// VHP-Flavored Markdown
//
//  * CVE identifiers link to the CVE database
//  * Icons inside of colons get converted to <i> tags.
//  * Markdown syntax (URLs are auto-converted to an <a> tag)
//  * Sanitize goes over the HTML code to remove anything XSS-related
//
export function vhpMarkdown(str) {
  str = linkifyCVE(str);
  str = embedVHPIcons(str);
  let converter = new showdown.Converter({
    simplifiedAutoLink: true, // find any URLs and link em
  });
  str = converter.makeHtml(str);
  str = cleanHtml(str);
  return str;
}

// Autolink all vulnerabilities following the format CVE-####-#### (case insensitive).
// The rightmost number can be 4+ digits long according to
// https://cve.mitre.org/about/faqs.html#cve_id_syntax_change
function linkifyCVE(str) {
  const regexp = /(CVE-\d{4}-\d{4,}?)/gi;
  const linkText = "[:vulnerability:$1](/cves/$1)";
  return str.replace(regexp, linkText);
}

// Identify all icon names placed between two colons, like :fi-camera:
// All icon names are replaced with an <i> tag and appropriate class.
// ":mi-bug-report:" turns into "<i class=\"vhp-icons-mi-bug-report\"></i>"
function embedVHPIcons(str) {
  const regex = /:(\w+?):/gi;
  const replaceText = '<i class="vhp-icon-$1"></i>';
  return str.replace(regex, replaceText);
}

function cleanHtml(str) {
  // use sanitize-html library here
  const opts = {
    allowedTags: [
      "p",
      "a",
      "ul",
      "ol",
      "li",
      "b",
      "i",
      "strong",
      "em",
      "strike",
      "abbr",
      "code",
      "hr",
      "pre",
      "h1",
      "h2",
      "h3",
    ],
    disallowedTagsMode: "discard",
    allowedSchemes: ["http", "https"],
    allowedAttributes: {
      a: ["href", "name", "target"],
      i: ["class"],
    },
  };
  return sanitizeHtml(str, opts);
}
