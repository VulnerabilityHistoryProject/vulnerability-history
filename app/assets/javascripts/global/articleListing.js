// import $ from 'jquery'
// import _ from 'lodash'
import {vhpMarkdown} from './vhpMarkdown'

export default function articleListing(jsonData, container = '#article-listing') {
  const blurbCounts = {
    '#blurb-1big-template' : 1,
    '#blurb-4even-template' : 4,
    '#blurb-1wide-template' : 1,
    '#blurb-1big1small-template' : 2,
  }
  const numTemplates = _.size(blurbCounts);
  const bMax = _.reduce(blurbCounts, (max, v, _k) => max = v > max ? v : max, 0)
  let articlesLeft = jsonData.length;
  let i = 0;
  while(articlesLeft > bMax){
    const templateID = Object.keys(blurbCounts)[i];
    addBlurbRow(container, templateID)
    i = (i + 1) % numTemplates;
    articlesLeft -= blurbCounts[templateID];
  }
  //for the last few we add a single wide row.
  _.times(articlesLeft, () => addBlurbRow(container, '#blurb-1wide-template'))

  $('.hidden').remove() //Remove leftover blurb templates

  let blurbs = $('.blurb-text')
  let blurbArts = $('.blurb-art')
  for (let i = 0; i < jsonData.length; i++) {
    let article = jsonData[i]
    $(blurbArts[i]).html(`<div class="vhp-art-${article.art}"/>`)
    blurbs[i].innerHTML = vhpMarkdown(`
### ${article.title} ###
${article.blurb}
`) + `<a href="/articles/${article.slug}" class="button read-more">Read more</a>`
  }
}

export function articleListingOneSize(jsonData, container = '#article-listing-same') {
  _.times(jsonData.length,() => addBlurbRow(container, '#blurb-1wide-template'))
  let blurbs = $(container + ' .blurb-text');
  let blurbArts = $(container + ' .blurb-art');
  for (let i = 0; i < jsonData.length; i++) {
    let article = jsonData[i];
    $(blurbArts[i]).html(`<div class="vhp-art-${article.art}"/>`);
    blurbs[i].innerHTML = vhpMarkdown(`
### ${article.title} ###
${article.blurb}
[Full Article](/articles/${article.slug})
    `);
  }
}

function addBlurbRow(container, templateID) {
  const blurbRow = $(templateID).clone();
  blurbRow.removeAttr('id').removeClass('hidden');
  $(container).append(blurbRow).append(`<hr/>`);
}
