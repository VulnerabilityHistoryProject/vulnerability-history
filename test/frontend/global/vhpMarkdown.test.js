import {vhpMarkdownAll, vhpMarkdown} from '../../../app/frontend/global/vhpMarkdown'


var linkString;
var iconString;
var xssString;
var heavyString;


const allDoc =
    '<div class="vhp-markdown">' +
        '<p>Heap-based buffer overflow in Chrome\'s resource provider in Google Chrome that' +
        'handles bitmap textures. The vulnerability allows remote attackers to cause a' +
        'denial of service or possibly have unspecified other impact via a large bitmap' +
        'texture size that triggers improper memory allocation in the software' +
        'renderer.' +
        '</p>' +
        '<p>The fix for this vulnerability is to place a hard limit on the number of bytes' +
        'that can be allocated in a specific situation, regardless of different kinds' +
        'of hardware.' +
        '</p>'+
    '</div>' +
    '<div class="tabs-panel is-active vhp-markdown" id="mistakes" role="tabpanel" aria-labelledby="mistakes-label">' +
        '<p>The vulnerabliity here was a pretty simple oversight in terms of limiting resource constraints. The code would have handled a massive amounts of memory allocation differently on different pieces of hardware,' +
        ' leading to a buffer overflow in some situations. They didn\'t think of this early on. But, the way they fixed it was to hardcode a reasonable limit into the code.' +
        'The use of a magic number without a constant still seems fishy to me, but it\'s very good that they considered this situation and rectified it. Interestingly, they did not trust the fuzzer on this initially.' +
        '</p>' +
    '</div>'

beforeEach(() => {
    linkString = "CVE-1111-2222";
    iconString = ":mi-bug-report:";
    xssString = "http://localhost:3000/<SCRIPT SRC=http://xss.rocks/xss.js></SCRIPT>";
    heavyString = linkString + iconString + xssString;
});

test("Mark Down", () => {
    expect(vhpMarkdown(linkString)).toBe("<p><a href=\"/cves/CVE-1111-2222\">\<i class=\"vhp-icon-vulnerability\"></i>CVE-1111-2222</a></p>");
    expect(vhpMarkdown(iconString)).toBe("<p>:mi-bug-report:</p>");
    expect(vhpMarkdown(xssString)).toBe("<p>http://localhost:3000/</p>");
    expect(vhpMarkdown(heavyString)).toBe("<p><a href=\"/cves/CVE-1111-2222\"><i class=\"vhp-icon-vulnerability\"></i>CVE-1111-2222</a>:mi-bug-report<i class=\"vhp-icon-http\"></i>//localhost:3000/</p>");
});

test("Mark Down All", () => {
    document.body.innerHTML = allDoc;
    vhpMarkdownAll();
    expect(document.body).toMatchSnapshot();
});
