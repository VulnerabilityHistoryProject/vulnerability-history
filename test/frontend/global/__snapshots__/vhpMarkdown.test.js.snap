// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mark Down All 1`] = `
<body>
  <div
    class="vhp-markdown"
  >
    <p>
      Heap-based buffer overflow in Chrome's resource provider in Google Chrome thathandles bitmap textures. The vulnerability allows remote attackers to cause adenial of service or possibly have unspecified other impact via a large bitmaptexture size that triggers improper memory allocation in the softwarerenderer.The fix for this vulnerability is to place a hard limit on the number of bytesthat can be allocated in a specific situation, regardless of different kindsof hardware.
    </p>
  </div>
  <div
    aria-labelledby="mistakes-label"
    class="tabs-panel is-active vhp-markdown"
    id="mistakes"
    role="tabpanel"
  >
    <p>
      The vulnerabliity here was a pretty simple oversight in terms of limiting resource constraints. The code would have handled a massive amounts of memory allocation differently on different pieces of hardware, leading to a buffer overflow in some situations. They didn't think of this early on. But, the way they fixed it was to hardcode a reasonable limit into the code.The use of a magic number without a constant still seems fishy to me, but it's very good that they considered this situation and rectified it. Interestingly, they did not trust the fuzzer on this initially.
    </p>
  </div>
</body>
`;
